gammapy/irf/core.py

```diff
@@ -1,17 +1,284 @@
+
+        This method uses log-log trapezoidal integration.
+
+        Parameters
+        ----------
+        **kwargs : dict
+            Coordinates at which to evaluate the IRF
+        method : {'linear', 'nearest'}, optional
+            Interpolation method
 
+        Returns
+        -------
+        array : `~astropy.units.Quantity`
+            Returns 2D array with axes offset
+        """
+        axis = self.axes.index("energy")
+        data = self.evaluate(**kwargs, method=method)
+        energy = kwargs["energy"]
+        return trapz_loglog(data, energy, axis=axis)
+
+    # TODO: define a proper integration method
+    @lazyproperty
+    def _integrate_rad(self):
+        rad_axis = self.axes["rad"]
+        rad_drad = (
+                2 * np.pi * rad_axis.center * self.quantity * rad_axis.bin_width
+        )
+        idx_rad = self.axes.index("rad")
+        values = rad_drad.cumsum(axis=idx_rad).to_value("")
+        values = np.insert(values, 0, 0, axis=idx_rad)
+
+        points = [ax.center for ax in self.axes]
+        points[idx_rad] = rad_axis.edges
+        return ScaledRegularGridInterpolator(
+            points=points, values=values, fill_value=1,
+        )
+
+    @classmethod
+    def from_hdulist(cls, hdulist, hdu=None):
+        """Create from `~astropy.io.fits.HDUList`.
+
+        Parameters
+        ----------
+        hdulist : `~astropy.io.HDUList`
+            HDU list
+        hdu : str
+            HDU name
+
+        Returns
+        -------
+        irf : `IRF`
+            IRF class
+        """
+        if hdu is None:
+            hdu = IRF_DL3_HDU_SPECIFICATION[cls.tag]["hdu"]
+
+        return cls.from_table(Table.read(hdulist[hdu]))
+
+    @classmethod
+    def read(cls, filename, hdu=None):
+        """Read from file.
+
+        Parameters
+        ----------
+        filename : str or `Path`
+            Filename
+        hdu : str
+            HDU name
+
+        Returns
+        -------
+        irf : `IRF`
+            IRF class
+        """
+        with fits.open(str(make_path(filename)), memmap=False) as hdulist:
+            return cls.from_hdulist(hdulist, hdu=hdu)
+
+    @classmethod
+    def from_table(cls, table):
+        """Read from `~astropy.table.Table`.
+
+        Parameters
+        ----------
+        table : `~astropy.table.Table`
+            Table with irf data
+
+        Returns
+        -------
+        irf : `IRF`
+            IRF class.
+        """
+        axes = MapAxes.from_table(table=table, format="gadf-dl3")[cls.required_axes]
+        column_name = IRF_DL3_HDU_SPECIFICATION[cls.tag]["column"]
+        data = table[column_name].quantity[0].transpose()
+
+        return cls(
+            axes=axes,
+            data=data.value,
+            meta=table.meta,
+            unit=data.unit
+        )
+
+    def to_table(self):
+        """Convert to `~astropy.table.Table`."""
+        table = self.axes.to_table(format="gadf-dl3")
```

Comment:
I think that in the long run having format specification inside `MapAxis` will lead to complexity in code duplication. 
Ideally we should have an `IRFReader` that knows about formats.

Reply:
I agree this will be the long term solution...